lib/elixir/lib/kernel.ex

```diff
@@ -4295,6 +4295,95 @@ defmodule Kernel do
   end
 
   @doc """
+  Generates a macro suitable for use in guard expressions.
+
+  It raises at compile time if the definition uses expressions that aren't
+  allowed in guards, and otherwise creates a macro that can be used both inside
+  or outside guards.
+
+  ## Example
+
+      defmodule Integer.Guards do
+        defguard is_even(value) when is_integer(value) and rem(value, 2) == 0
+      end
+
+      defmodule Collatz do
+        @moduledoc "Tools for working with the Collatz sequence."
+        import Integer.Guards
+
+        @doc "Determines the number of steps `n` takes to reach `1`."
+        # If this function never converges, please let me know what `n` you used.
+        def converge(n) when n > 0, do: step(n, 0)
+
+        defp step(1, step_count) do
+          step_count
+        end
+        defp step(n, step_count) when is_even(n) do
+          step(div(n, 2), step_count + 1)
+        end
+        defp step(n, step_count) do
+          step(3*n + 1, step_count + 1)
+        end
+      end
+
+  """
+  @spec defguard(Macro.t) :: Macro.t | no_return
+  defmacro defguard(guard) do
+    define_guard(:defmacro, guard, __CALLER__)
+  end
+
+  @doc """
+  Generates a private macro suitable for use in guard expressions.
+
+  It raises at compile time if the definition uses expressions that aren't
+  allowed in guards, and otherwise creates a private macro that can be used both inside
+  or outside guards.
+
+  """
+  @spec defguardp(Macro.t) :: Macro.t | no_return
+  defmacro defguardp(guard) do
+    define_guard(:defmacrop, guard, __CALLER__)
+  end
+
+  defp define_guard(kind, guard, env) do
+    case :elixir_utils.extract_guards(guard) do
+      {call, impl} when length(impl) < 2 ->
+        case Macro.decompose_call(call) do
+          {_name, args} ->
+            validate_variable_only_args!(call, args)
+            {^args, refs} = extract_refs_from_args(args)
```

Comment:
This and the line above could be done in one pass now. One that validates and extracts.